Potential Deadlock when jsonrpc2.Logger sends messages over same jsonrpc2.Conn

When using this library, we had the use case of forwarding log messages back to the other peer, using the `jsonrpc2.SetLogger` `jsonrpc2.ConnOpt`. This, however, might result in an internal deadlock.

I am unsure how much of a bug this is, but the API allows it and I have not found any warning in the docs. If this is unsupported behavior, please consider adding an API doc comment stating so.

Thanks!

## Trace

When the underlying `Conn` is closed with a known cause, the logger is called while the internal mutex is locked.

https://github.com/sourcegraph/jsonrpc2/blob/2cc94179e136b565cb9df7d9d0bff9fc06912ac9/conn.go#L169-L182

Now, `c.logger.Printf` results in `e.conn.Notify` being called, which calls `c.send`.

https://github.com/sourcegraph/jsonrpc2/blob/2cc94179e136b565cb9df7d9d0bff9fc06912ac9/conn.go#L127-L142

And there, another mutex lock is being set, while the prior lock is still active.

https://github.com/sourcegraph/jsonrpc2/blob/2cc94179e136b565cb9df7d9d0bff9fc06912ac9/conn.go#L241-L260

Btw, `Conn.close` is only called with a cause in `c.readMessage` if the next JSON RPC message cannot be read.

https://github.com/sourcegraph/jsonrpc2/blob/2cc94179e136b565cb9df7d9d0bff9fc06912ac9/conn.go#L191-L194

## Proof of Concept

```go
package main

import (
	"context"
	"errors"
	"io"
	"runtime/debug"
	"time"

	"github.com/sourcegraph/jsonrpc2"
)

// logger using the underlying jsonrpc2.Conn. The pointer-pointer is used to ease self-reference in main below.
type logger struct{ conn **jsonrpc2.Conn }

func (l logger) Printf(string, ...any) {
	_ = (*l.conn).Notify(context.Background(), "log", nil)
}

type handler struct{}

func (handler) Handle(context.Context, *jsonrpc2.Conn, *jsonrpc2.Request) {}

type readWriteCloser struct {
	r io.ReadCloser
	w io.WriteCloser
}

func (rwc *readWriteCloser) Read(p []byte) (n int, err error)  { return rwc.r.Read(p) }
func (rwc *readWriteCloser) Write(p []byte) (n int, err error) { return rwc.w.Write(p) }
func (rwc *readWriteCloser) Close() error                      { return errors.Join(rwc.r.Close(), rwc.w.Close()) }

func main() {
	debug.SetTraceback("all")

	r, w := io.Pipe()

	var conn *jsonrpc2.Conn
	conn = jsonrpc2.NewConn(
		context.Background(),
		jsonrpc2.NewPlainObjectStream(&readWriteCloser{r, w}),
		handler{},
		jsonrpc2.SetLogger(logger{&conn}),
	)

	// Faulty or malicious peer sends invalid JSON RPC message.
	_, _ = w.Write([]byte("☃\n"))

	select {
	case <-conn.DisconnectNotify():
	case <-time.After(3 * time.Second):
		panic("timeout")
	}
}
```

```
$ go run main.go
panic: timeout

goroutine 1 [running]:
main.main()
        …/main.go:52 +0x485

goroutine 7 [sync.Mutex.Lock]:
internal/sync.runtime_SemacquireMutex(0x1bb09bbebc08?, 0x25?, 0x1bb09bbebc18?)
        /nix/store/rcz9i4msbg3178grqll9h98b07dwk7zg-go-1.26.5/share/go/src/runtime/sema.go:95 +0x25
internal/sync.(*Mutex).lockSlow(0x1bb09bbf8020)
        /nix/store/rcz9i4msbg3178grqll9h98b07dwk7zg-go-1.26.5/share/go/src/internal/sync/mutex.go:149 +0x15d
internal/sync.(*Mutex).Lock(...)
        /nix/store/rcz9i4msbg3178grqll9h98b07dwk7zg-go-1.26.5/share/go/src/internal/sync/mutex.go:70
sync.(*Mutex).Lock(...)
        /nix/store/rcz9i4msbg3178grqll9h98b07dwk7zg-go-1.26.5/share/go/src/sync/mutex.go:46
github.com/sourcegraph/jsonrpc2.(*Conn).send(0x1bb09bbf8000, {0x1bb09bbebd78?, 0x4?}, 0x1bb09bb520d0, 0x0)
        …/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.2.1/conn.go:260 +0x12d
github.com/sourcegraph/jsonrpc2.(*Conn).Notify(0x1bb09bbf8000, {0x529988, 0x63a100}, {0x51c18f, 0x3}, {0x0, 0x0}, {0x0, 0x0, 0x0})
        …/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.2.1/conn.go:142 +0x1ce
main.logger.Printf({0x4?}, {0x1bb09bb4e060?, 0x1bb09bbebe58?}, {0x1bb09bbebe78?, 0x4d84ac?, 0x1bb09bbf60c0?})
        …/go/jsonrpc2-poc-zhjl/main.go:17 +0x45
github.com/sourcegraph/jsonrpc2.(*Conn).close(0x1bb09bbf8000, {0x529660, 0x1bb09bb4e060})
        …/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.2.1/conn.go:181 +0x257
github.com/sourcegraph/jsonrpc2.(*Conn).readMessages(0x1bb09bbf8000, {0x529988, 0x63a100})
        …/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.2.1/conn.go:194 +0x2c6
created by github.com/sourcegraph/jsonrpc2.NewConn in goroutine 1
        /…/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.2.1/conn.go:62 +0x1e6
exit status 2
```

--- a/conn.go
+++ b/conn.go
@@ -180,24 +180,29 @@ func (c *Conn) SendResponse(ctx context.Context, resp *Response) error {
 
 func (c *Conn) close(cause error) error {
 	c.mu.Lock()
-	defer c.mu.Unlock()
-
 	if c.closed {
+		c.mu.Unlock()
 		return ErrClosed
 	}
+	c.closed = true
 
 	for _, call := range c.pending {
 		close(call.done)
 	}
 
+	close(c.disconnect)
+	c.mu.Unlock()
+
+	c.cancelCtx()
+	err := c.stream.Close()
+
+	// The logger may call back into c, so invoke it only after shutdown is
+	// complete and c.mu is unlocked.
 	if cause != nil && cause != io.EOF && cause != io.ErrUnexpectedEOF {
 		c.logger.Printf("jsonrpc2: protocol error: %v\n", cause)
 	}
 
-	close(c.disconnect)
-	c.cancelCtx()
-	c.closed = true
-	return c.stream.Close()
+	return err
 }
 
 func (c *Conn) readMessages(ctx context.Context) {

--- a/conn_test.go
+++ b/conn_test.go
@@ -167,6 +167,38 @@ func TestConn_DisconnectNotify(t *testing.T) {
 		connA.Write([]byte("invalid json"))
 		assertDisconnect(t, c, connB)
 	})
+
+	t.Run("protocol error logger uses connection", func(t *testing.T) {
+		connA, connB := net.Pipe()
+		logResult := make(chan error, 1)
+		var c *jsonrpc2.Conn
+		c = jsonrpc2.NewConn(
+			context.Background(),
+			jsonrpc2.NewPlainObjectStream(connB),
+			noopHandler{},
+			jsonrpc2.SetLogger(connNotifyLogger{conn: &c, result: logResult}),
+		)
+		connA.Write([]byte("invalid json"))
+		assertDisconnect(t, c, connB)
+
+		select {
+		case err := <-logResult:
+			if err != jsonrpc2.ErrClosed {
+				t.Errorf("logger Notify: got %v, want %v", err, jsonrpc2.ErrClosed)
+			}
+		case <-time.After(200 * time.Millisecond):
+			t.Error("logger blocked using connection")
+		}
+	})
+}
+
+type connNotifyLogger struct {
+	conn   **jsonrpc2.Conn
+	result chan<- error
+}
+
+func (l connNotifyLogger) Printf(string, ...interface{}) {
+	l.result <- (*l.conn).Notify(context.Background(), "log", nil)
 }
 
 func TestConn_Close(t *testing.T) {

